### ISSUE TITLE
errno=22 when submitting an usb_transfer with wLength larger than the data payload 

### ISSUE THREAD

**Issue by CestCarre** [libusb/libusb#676]
Hi everyone,

I'm facing an issue with libusb, maybe someone knows what's going wrong ?

I'm trying to submit an usb_transfer with a wLength larger than the data payload.
Here is the C code:

```c
int main() {
  int ret = 0;
  init_usb(); // init usb, nothing special here
  libusb_device_handle* handler = NULL;
  libusb_set_debug(LibUsbCtx, 4);
  handler = libusb_open_device_with_vid_pid(LibUsbCtx, VID, PID);
  if (handler == NULL) {
    puts("No device found.");
    ret = 1;
    goto out;
  }
  libusb_detach_kernel_driver(handler, 0);
  if (libusb_claim_interface(handler, 0) < 0) {
    puts("Unable to claim interface");
    ret = 2;
    goto out;
  }

  uint8_t buffer[8 + 0x40];
  memset((void*) buffer, 0, sizeof(buffer));
  buffer[0] = 0x21; // uint8_t
  buffer[1] = 3; // uint8_t
  buffer[2] = 0; // uint16_t
  buffer[3] = 0;
  buffer[4] = 0; // uint16_t
  buffer[5] = 0;
  buffer[6] = 0x00; // uint16_t
  buffer[7] = 0x08;                   // Here, I set wLength to 0x800, but the data size is 0x40
                                               // If I set wLength to 0x40, everything works perfectly.
  memset((void*) (buffer + 8), '\x00', 0x40);

  struct libusb_transfer* tr = libusb_alloc_transfer(0);
  tr->dev_handle = handler;
  tr->endpoint = 0;
  tr->type = 0;
  tr->timeout = 0;
  tr->buffer = buffer;
  tr->length = sizeof(buffer); // So 0x40 + 8
  tr->actual_length = 0;
  tr->user_data = NULL;
  tr->callback = callback;
  tr->flags = 0;

  puts("Send packet");
  int ret_submit = libusb_submit_transfer(tr);
  printf("ret=%d\n", ret_submit);

  while (!cont) {
    int r = libusb_handle_events(LibUsbCtx);
    if (r < 0) {
      printf("Error: %s\n", libusb_strerror(r));
      cont = 1;
    }
  }

out:
  deinit_usb();
  return ret;
}
```

Here is the output from the term:
```
[timestamp] [threadID] facility level [function call] <message>
--------------------------------------------------------------------------------
[ 0.010047] [00000f95] libusb: debug [libusb_get_device_list]
[ 0.010101] [00000f95] libusb: debug [libusb_get_device_descriptor]
[ 0.010109] [00000f95] libusb: debug [libusb_get_device_descriptor]
[ 0.010124] [00000f95] libusb: debug [libusb_get_device_descriptor]
[ 0.010140] [00000f95] libusb: debug [libusb_get_device_descriptor]
[ 0.010153] [00000f95] libusb: debug [libusb_get_device_descriptor]
[ 0.010167] [00000f95] libusb: debug [libusb_open] open 1.54
[ 0.010216] [00000f95] libusb: debug [usbi_add_pollfd] add fd 9 events 4
[ 0.010231] [00000f95] libusb: debug [libusb_detach_kernel_driver] interface 0
[ 0.010253] [00000f95] libusb: debug [libusb_claim_interface] interface 0
[ 0.010363] [00000f95] libusb: debug [libusb_alloc_transfer] transfer 0x1144b50
Send packer
[ 0.010410] [00000f95] libusb: debug [libusb_submit_transfer] transfer 0x1144b50
[ 0.010427] [00000f95] libusb: error [submit_control_transfer] submiturb failed error -1 errno=22
ret=-1
[ 0.010455] [00000f95] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.010471] [00000f95] libusb: debug [handle_events] poll fds modified, reallocating
[ 0.010488] [00000f95] libusb: debug [handle_events] poll() 3 fds with timeout in 60000ms
```

Is it legal to do so ?

**Comment by dickens** [comment]
No, it is not legal to do so.  The wLength tells the kernel how much space it has available to read or write data for the transfer.  In your case if the kernel didn’t check against the transfer length it would have read 0x800 bytes (note that the wLength in your example is _0x800_ and not 0x80) starting at `buffer + 8`, reading 1,984 bytes of garbage or whatever happens to be adjacent to `buffer`.

In other words, `transfer->length` must be greater than or equal to `wLength + 8`.
